File: frontend/src/features/activity/activityRepository.ts
import { ApiHttpError, fetchJson } from '../../lib/api/client'
import type { ActualBlock, TaskType } from '../../lib/api'

export const activityDevelopmentEnabled = (import.meta.env.DEV || import.meta.env.MODE === 'activity-review') && import.meta.env.VITE_ACTIVITY_TRACKING_DEV === '1'
export interface ActivityPlan {
  id: number; task_type_id: number; task_id: number | null; name: string | null; note: string | null; start_at: string; end_at: string
}
export interface ActivitySelection { task_id?: number | null; planned_block_id?: number | null; note?: string | null }
export interface ActivitySnapshot {
  protocol: 'activity-online-v1'; cursor: number; server_at: string; reporting_timezone: string
  current: ActualBlock | null; records: ActualBlock[]; task_types?: TaskType[]
  reporting_timezone_initialized?: boolean
  plans?: ActivityPlan[]
  offline_ready?: boolean
  operation_outcomes?: Record<string, { device_id: string; outcome: string }>
  coverage?: { start: string; end: string | null; record_id: number | null; order: [string, string, number, string] }[]
  acknowledgement?: { operation_id: string; outcome: string } | null
}
interface Command {
  operation_id: string; device_id: string; sequence: number; action_at: string
  calibration: { server_at: string; offset_ms: number }; base_cursor: number
  effective: { mode: 'server_now' | 'instant'; at?: string }; target_id: number | null
  selection_snapshot?: boolean; task_id?: number | null; planned_block_id?: number | null; note?: string | null
  predecessor_id?: string; kind: 'start' | 'switch' | 'stop'; task_type_id?: number; name?: string
}
interface Journal {
  device: string; sequence: number; lastAction: number; pending: Command | null
  outbox: Command[]; rejected?: Command[] | Command | null; snapshot: ActivitySnapshot | null
  calibration?: Command['calibration']
}
const storageKey = `timebox.activity.online.v1:${import.meta.env.VITE_API_BASE_URL ?? '/api'}`
type Exclusive = <T>(work: () => Promise<T>) => Promise<T>
const browserExclusive: Exclusive = (work) => {
  if (!navigator.locks) return Promise.reject(new Error('Activity tracking requires browser storage locks. Open localhost or HTTPS.'))
  return navigator.locks.request(storageKey, work)
}
export class ActivityRepository {
  private listeners = new Set<() => void>()
  private journal: Journal = { device: crypto.randomUUID(), sequence: 0, lastAction: 0, pending: null, outbox: [], snapshot: null }
  private anchor: { server: number; monotonic: number } | null = null
  private offline = false
  private storageError: string | null = null
  private feedback: string | null = null
  state: { snapshot: ActivitySnapshot | null; pending: boolean; busy: boolean; error: string | null; offline: boolean; feedback?: string | null }
  private storage: Storage
  private exclusive: Exclusive
  constructor(storage: Storage, exclusive = browserExclusive) {
    this.storage = storage
    this.exclusive = exclusive
    try { this.journal = this.readJournal() } catch { this.storageError = 'Activity storage is unavailable. Recording has not started.' }
    this.state = { snapshot: this.project(), pending: this.journal.outbox.length > 0, busy: false, error: this.storageError, offline: false }
  }
  subscribe = (listener: () => void) => { this.listeners.add(listener); return () => { this.listeners.delete(listener) } }
  getSnapshot = () => this.state
  now = () => this.anchor ? this.anchor.server + performance.now() - this.anchor.monotonic : Date.now() + (this.journal.calibration?.offset_ms ?? 0)
  private project(): ActivitySnapshot | null {
    if (!this.journal.snapshot) return null
    const snapshot = { ...this.journal.snapshot, records: [...this.journal.snapshot.records] }
    if (snapshot.coverage?.length && this.journal.outbox.length) return this.projectRanges(snapshot)
    for (const command of this.journal.outbox) {
      // Old online receipts have no client effective instant; retain the confirmed view until replay.
      if (command.effective.mode !== 'instant') continue
      const at = command.effective.at!
      if (snapshot.current) snapshot.records = snapshot.records.map(row => row.id === snapshot.current!.id ? { ...row, end_at: at } : row)
      snapshot.current = null
      if (command.kind !== 'stop') {
        const type = snapshot.task_types?.find(t => t.id === command.task_type_id) ?? { id: command.task_type_id ?? 0, name: command.task_type_id ? 'Activity' : 'unspecified', created_at: at, updated_at: at }
        snapshot.current = { id: -command.sequence, task_type_id: type.id, task_type: type, task_id: command.task_id ?? null, task: null, name: command.name ?? null, note: command.note ?? null, planned_block_id: command.planned_block_id ?? null, start_at: at, end_at: null, created_at: at, updated_at: at }
        snapshot.records.push(snapshot.current)
      }
    }
    return snapshot
  }
  private projectRanges(snapshot: ActivitySnapshot): ActivitySnapshot {
    // Keep the API's microsecond precision when comparing canonical coverage.
    // Date alone truncates it and could turn different actions into a false tie.
    const infinity = 8640000000000000000n
    const instant = (value: string) => BigInt(Date.parse(value)) * 1000n + BigInt((value.match(/\.(\d+)/)?.[1] ?? '').padEnd(6, '0').slice(3, 6))
    const format = (value: bigint) => {
      const millis = value >= 0n ? value / 1000n : (value - 999n) / 1000n
      const remainder = value - millis * 1000n
      const date = new Date(Number(millis)).toISOString()
      return remainder === 0n ? date : date.replace('Z', `${String(remainder).padStart(3, '0')}Z`)
    }
    type Piece = { start: bigint; end: bigint; row: ActualBlock | null; order: [bigint, string, number, string] }
    const compare = (a: Piece['order'], b: Piece['order']) => {
      for (let i = 0; i < 4; i++) { if (a[i] < b[i]) return -1; if (a[i] > b[i]) return 1 }
      return 0
    }
    let pieces: Piece[] = snapshot.coverage!.map(p => ({ start: instant(p.start), end: p.end ? instant(p.end) : infinity,
      row: snapshot.records.find(r => r.id === p.record_id) ?? null,
      order: [p.order[0] ? instant(p.order[0]) : -infinity, p.order[1], p.order[2], p.order[3]] }))
    for (const command of this.journal.outbox) {
      if (command.effective.mode !== 'instant') continue
      const at = command.effective.at!
      const start = instant(at)
      const order: Piece['order'] = [instant(command.action_at), command.device_id, command.sequence, command.operation_id]
      const type = snapshot.task_types?.find(t => t.id === command.task_type_id) ?? { id: command.task_type_id ?? 0, name: 'unspecified', created_at: at, updated_at: at }
      const row: ActualBlock | null = command.kind === 'stop' ? null : { id: -command.sequence, task_type_id: type.id, task_type: type, task_id: command.task_id ?? null, task: null, name: command.name ?? null, note: command.note ?? null, planned_block_id: command.planned_block_id ?? null, start_at: at, end_at: null, created_at: at, updated_at: at }
      const boundaries = [...new Set([start, ...pieces.flatMap(p => [p.start, p.end])])].sort((a, b) => a < b ? -1 : a > b ? 1 : 0)
      if (boundaries.at(-1) !== infinity) boundaries.push(infinity)
      const next: Piece[] = []
      for (let i = 0; i < boundaries.length - 1; i++) {
        const a = boundaries[i], b = boundaries[i + 1]
        const previous = pieces.find(p => p.start <= a && p.end > a)
        const winner = a >= start && (!previous || compare(order, previous.order) > 0) ? { row, order } : previous
        if (winner) next.push({ start: a, end: b, row: winner.row, order: winner.order })
      }
      pieces = next
    }
    const merged: Piece[] = []
    for (const piece of pieces) {
      const previous = merged.at(-1)
      if (previous && previous.end === piece.start && previous.row?.id === piece.row?.id && compare(previous.order, piece.order) === 0) previous.end = piece.end
      else merged.push({ ...piece })
    }
    const records = merged.filter(p => p.row).map(p => ({ ...p.row!, start_at: format(p.start), end_at: p.end === infinity ? null : format(p.end) }))
    return { ...snapshot, records, current: records.find(r => r.end_at === null) ?? null }
  }
  private publish(error: string | null = null, busy = false) {
    this.state = { snapshot: this.project(), pending: this.journal.outbox.length > 0, error, busy, offline: this.offline, feedback: this.feedback }
    this.listeners.forEach(listener => listener())
  }
  private readJournal(): Journal {
    const raw = this.storage.getItem(storageKey)
    if (!raw) return { device: crypto.randomUUID(), sequence: 0, lastAction: 0, pending: null, outbox: [], snapshot: null }
    const value = JSON.parse(raw) as Journal
    return { ...value, outbox: value.outbox ?? (value.pending ? [value.pending] : []), pending: null }
  }
  private save(journal: Journal) {
    try { this.storage.setItem(storageKey, JSON.stringify(journal)) } catch { throw new Error('Activity storage failed. Change was not saved on this device.') }
    this.journal = journal
  }
  private newer(snapshot: ActivitySnapshot) {
    if (snapshot.protocol !== 'activity-online-v1') throw new Error('Incompatible activity server')
    const previous = this.journal.snapshot
    return !previous || snapshot.cursor > previous.cursor || (snapshot.cursor === previous.cursor && Date.parse(snapshot.server_at) >= Date.parse(previous.server_at))
  }
  dismissFeedback = () => { this.feedback = null; this.publish(this.state.error) }
  private noteReconciliation(snapshot: ActivitySnapshot) {
    if (!this.newer(snapshot)) return
    if (Object.entries(snapshot.operation_outcomes ?? {}).some(([id, result]) =>
      result.device_id === this.journal.device && result.outcome === 'superseded' &&
      this.journal.snapshot?.operation_outcomes?.[id]?.outcome !== 'superseded')) {
      this.feedback = 'A newer change on another device updated this time.'
    }
  }
  private async drain() {
    while (this.journal.outbox.length) {
      const command = this.journal.outbox[0]
      let response: ActivitySnapshot
      try { response = await fetchJson<ActivitySnapshot>('/activity/commands', { method: 'POST', body: JSON.stringify(command) }) }
      catch (error) {
        if (error instanceof ApiHttpError && error.status === 422) {
          this.save({ ...this.journal, rejected: this.journal.outbox, outbox: [] })
        } else this.offline = true
        throw error
      }
      if (response.acknowledgement?.operation_id !== command.operation_id) throw new Error('Activity acknowledgement is missing')
      const applied = ['applied', 'superseded'].includes(response.acknowledgement.outcome)
      this.noteReconciliation(response)
      // Receipt removal and canonical state advance are one durable write.
      this.save({ ...this.journal, snapshot: this.newer(response) ? response : this.journal.snapshot,
        outbox: applied ? this.journal.outbox.slice(1) : [], rejected: applied ? this.journal.rejected : this.journal.outbox })
      this.offline = false
      if (!applied) throw new Error('Activity changed on another device. Pending changes were retained for review.')
    }
  }
  async refresh() {
    try {
      await this.exclusive(async () => {
        if (this.storageError) throw new Error(this.storageError)
        this.journal = this.readJournal()
        // Replay before reading, avoiding a transient rollback of pending intent.
        await this.drain()
      })
      let snapshot: ActivitySnapshot
      try {
        snapshot = await fetchJson<ActivitySnapshot>('/activity')
        if (snapshot.reporting_timezone_initialized === false) snapshot = await fetchJson<ActivitySnapshot>('/activity/reporting-timezone/initialize', { method: 'POST', body: JSON.stringify({ timezone: Intl.DateTimeFormat().resolvedOptions().timeZone }) })
      } catch (error) { this.offline = true; throw error }
      await this.exclusive(async () => {
        this.journal = this.readJournal()
        this.offline = false
        if (this.newer(snapshot)) {
          this.noteReconciliation(snapshot)
          const server = Date.parse(snapshot.server_at)
          this.save({ ...this.journal, snapshot, calibration: { server_at: snapshot.server_at, offset_ms: server - Date.now() } })
          this.anchor = { server, monotonic: performance.now() }
        }
        this.publish()
      })
    } catch (error) { this.publish(error instanceof Error ? error.message : 'Could not refresh activity') }
  }
  async setReportingTimezone(timezone: string) {
    try {
      await this.exclusive(async () => {
        this.journal = this.readJournal()
        const snapshot = await fetchJson<ActivitySnapshot>('/activity/reporting-timezone', { method: 'PUT', body: JSON.stringify({ timezone }) })
        if (this.newer(snapshot)) this.save({ ...this.journal, snapshot })
        this.publish()
      })
      return true
    } catch (error) { this.publish(error instanceof Error ? error.message : 'Could not save time zone'); return false }
  }
  async command(kind: Command['kind'], taskTypeId?: number, name?: string, retryOnly = false, selection?: ActivitySelection) {
    if (retryOnly) { await this.refresh(); return !this.state.pending && !this.state.error }
    let saved = false
    const requestedAt = this.now()
    if (kind === 'start' && !selection && taskTypeId == null && !name) {
      const plan = this.currentPlan()
      if (plan) { taskTypeId = plan.task_type_id; name = plan.name ?? undefined; selection = { task_id: plan.task_id, planned_block_id: plan.id, note: plan.note } }
    }
    const observed = { snapshot: this.journal.snapshot, calibration: this.journal.calibration,
      current: this.project()?.current, predecessor: this.journal.outbox.at(-1) }
    try {
      await this.exclusive(async () => {
        if (this.storageError) throw new Error(this.storageError)
        this.journal = this.readJournal()
        if (!this.journal.snapshot?.offline_ready || !this.journal.calibration) throw new Error('Connect once to an updated server to initialize Activity Tracking before recording offline.')
        if (this.journal.outbox.some(command => command.effective.mode === 'server_now')) throw new Error('Reconnect to confirm the previous online change first.')
        const projected = this.project()!
        if ((kind === 'start') === !!projected.current) throw new Error('Activity changed. Review the current activity.')
        if (kind === 'switch' && !taskTypeId && !selection?.task_id) throw new Error('Task Type is required')
        const latest = projected.records.reduce((value, row) => Math.max(value, Date.parse(row.end_at ?? row.start_at)), 0)
        const action = Math.max(this.journal.lastAction + 1, latest + 1, requestedAt)
        const at = new Date(action).toISOString()
        const predecessor = this.journal.outbox.at(-1) ?? observed.predecessor
        const command: Command = { operation_id: crypto.randomUUID(), device_id: this.journal.device, sequence: this.journal.sequence + 1,
          action_at: at, calibration: observed.calibration ?? this.journal.calibration, base_cursor: observed.snapshot?.cursor ?? this.journal.snapshot.cursor,
          effective: { mode: 'instant', at }, target_id: predecessor ? null : observed.current?.id ?? null,
          ...(predecessor ? { predecessor_id: predecessor.operation_id } : {}), kind,
          selection_snapshot: true, ...selection,
          ...(taskTypeId == null ? {} : { task_type_id: taskTypeId }), ...(name ? { name } : {}) }
        this.save({ ...this.journal, sequence: command.sequence, lastAction: action, outbox: [...this.journal.outbox, command] })
        saved = true
        this.publish()
      })
    } catch (error) { this.publish(error instanceof Error ? error.message : 'Could not save activity'); return false }
    if (saved) void this.refresh()
    return saved
  }
  currentPlan = () => this.state.snapshot?.plans?.find(p => Date.parse(p.start_at) <= this.now() && this.now() < Date.parse(p.end_at))
  async trackTask(task: { id: number; title: string; task_type_id: number | null; recurrence_kind?: string | null; status: string }) {
    if (task.recurrence_kind === 'quota_parent' || task.status === 'completed') return false
    if (!this.state.snapshot) await this.refresh()
    const type = task.task_type_id ?? this.state.snapshot?.task_types?.find(t => t.name === 'unspecified')?.id
    const result = await this.command(this.state.snapshot?.current ? 'switch' : 'start', type, task.title, false, { task_id: task.id })
    if (result) { this.feedback = `Tracking ${task.title}`; this.publish(this.state.error) }
    return result
  }
  adoptPlan = (plan: ActivityPlan) => this.command(this.state.snapshot?.current ? 'switch' : 'start', plan.task_type_id, plan.name ?? undefined, false, { task_id: plan.task_id, planned_block_id: plan.id, note: plan.note })
  retry = () => this.refresh()
}
let shared: ActivityRepository | undefined
export const getActivityRepository = () => shared ??= new ActivityRepository(window.localStorage)

File: android/app/src/main/java/com/timebox/android/data/ActivityRepository.kt
package com.timebox.android.data

import android.content.Context
import com.timebox.android.data.remote.ActivityCalibrationDto
import com.timebox.android.data.remote.ActivityCommandDto
import com.timebox.android.data.remote.ActivityEffectiveDto
import com.timebox.android.data.remote.ActivityKind
import com.timebox.android.data.remote.ActivityOutcome
import com.timebox.android.data.remote.ActivitySnapshotDto
import com.timebox.android.data.remote.ApiFactory
import java.time.Instant
import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString

interface ActivityTransport {
    suspend fun read(): ActivitySnapshotDto
    suspend fun execute(command: ActivityCommandDto): ActivitySnapshotDto
    suspend fun endpoint(): String = "test"
}
class RepositoryActivityTransport(private val repository: TimeboxRepository) : ActivityTransport {
    override suspend fun read() = repository.getActivity()
    override suspend fun execute(command: ActivityCommandDto) = repository.activityCommand(command)
    override suspend fun endpoint() = repository.activityEndpoint()
}
interface ActivityStorage {
    fun load(): String?
    fun save(value: String)
}
class AndroidActivityStorage(context: Context) : ActivityStorage {
    private val preferences = context.getSharedPreferences("activity-online-v1", Context.MODE_PRIVATE)
    override fun load() = preferences.getString("journal", null)
    override fun save(value: String) {
        check(preferences.edit().putString("journal", value).commit()) { "Activity storage failed. Change has not been confirmed." }
    }
}
@Serializable private data class ActivityJournal(
    val device: String = UUID.randomUUID().toString(), val sequence: Int = 0,
    val lastAction: Long = 0, val endpoint: String? = null,
    val pending: ActivityCommandDto? = null, val snapshot: ActivitySnapshotDto? = null,
    val rejected: ActivityCommandDto? = null,
)
data class ActivityUiState(
    val snapshot: ActivitySnapshotDto? = null, val pending: Boolean = false,
    val busy: Boolean = false, val error: String? = null,
)

/** A durable online receipt/retry boundary, independent of navigation and Work Mode. */
class ActivityRepository(private val transport: ActivityTransport, private val storage: ActivityStorage) {
    private val mutex = Mutex()
    private var journal = ActivityJournal()
    private val mutableState = MutableStateFlow(ActivityUiState())
    val state = mutableState.asStateFlow()
    private var serverAnchor: Long? = null
    private var monotonicAnchor = 0L
    private var calibration: ActivityCalibrationDto? = null
    private var storageError: String? = null
    init {
        try {
            storage.load()?.let { journal = ApiFactory.json.decodeFromString<ActivityJournal>(it) }
            publish()
        } catch (error: Exception) {
            storageError = "Activity storage unavailable: ${error.message}"
            publish(storageError)
        }
    }
    private fun publish(error: String? = null, busy: Boolean = false) {
        mutableState.value = ActivityUiState(journal.snapshot, journal.pending != null, busy, error)
    }
    private fun save(value: ActivityJournal) {
        storage.save(ApiFactory.json.encodeToString(value))
        journal = value
    }
    private suspend fun checkEndpoint() {
        check(storageError == null) { storageError!! }
        val endpoint = transport.endpoint()
        check(journal.endpoint == null || journal.endpoint == endpoint) { "Return to the original activity server to recover this device's recording." }
        if (journal.endpoint == null) save(journal.copy(endpoint = endpoint))
    }
    private fun accept(snapshot: ActivitySnapshotDto) {
        check(snapshot.protocol == "activity-online-v1") { "Incompatible activity server" }
        if (snapshot.cursor >= (journal.snapshot?.cursor ?: -1)) save(journal.copy(snapshot = snapshot))
    }
    suspend fun refresh() = mutex.withLock {
        try {
            checkEndpoint()
            val snapshot = transport.read()
            accept(snapshot)
            serverAnchor = Instant.parse(snapshot.serverAt).toEpochMilli()
            monotonicAnchor = System.nanoTime()
            calibration = ActivityCalibrationDto(snapshot.serverAt, serverAnchor!! - System.currentTimeMillis())
            publish(if (journal.pending != null) "Change not confirmed. Retry to check the saved result." else null)
        } catch (cancelled: CancellationException) { throw cancelled
        } catch (error: Exception) { publish(error.message ?: "Could not refresh activity") }
    }
    fun now(): Instant = serverAnchor?.let { Instant.ofEpochMilli(it + (System.nanoTime() - monotonicAnchor) / 1_000_000) } ?: Instant.now()
    suspend fun command(kind: ActivityKind, taskTypeId: Int? = null, name: String? = null, retryOnly: Boolean = false): Boolean = mutex.withLock {
        publish(busy = true)
        try {
            checkEndpoint()
            if (retryOnly && journal.pending == null) { publish(); return@withLock false }
            check(retryOnly || journal.pending == null) { "Retry the unconfirmed change before recording another activity" }
            if (journal.pending == null) {
                val snapshot = checkNotNull(journal.snapshot) { "Refresh before recording" }
                val server = checkNotNull(serverAnchor) { "Refresh before recording" }
                val action = maxOf(journal.lastAction, server + (System.nanoTime() - monotonicAnchor) / 1_000_000)
                val sequence = journal.sequence + 1
                val pending = ActivityCommandDto(UUID.randomUUID().toString(), journal.device, sequence,
                    Instant.ofEpochMilli(action).toString(), checkNotNull(calibration), snapshot.cursor,
                    ActivityEffectiveDto("server_now"), snapshot.current?.id, kind, taskTypeId, name?.trim()?.ifEmpty { null })
                save(journal.copy(sequence = sequence, lastAction = action, pending = pending))
            }
            val pending = checkNotNull(journal.pending)
            val response = try { transport.execute(pending) } catch (error: retrofit2.HttpException) {
                if (error.code() == 422) save(journal.copy(rejected = pending, pending = null))
                throw error
            }
            val ack = checkNotNull(response.acknowledgement) { "Activity acknowledgement missing" }
            check(ack.operationId == pending.operationId) { "Activity acknowledgement does not match" }
            accept(response)
            save(journal.copy(pending = null))
            publish(if (ack.outcome == ActivityOutcome.Conflict) "Activity changed on another device. Review before trying again." else null)
            ack.outcome == ActivityOutcome.Applied
        } catch (cancelled: CancellationException) { publish("Change not confirmed. Retry to check the saved result."); throw cancelled
        } catch (error: Exception) { publish(error.message ?: "Change not confirmed"); false }
    }
    suspend fun retry() = command(journal.pending?.kind ?: ActivityKind.Start, retryOnly = true)
}
